content/blog/2025-04-09-introducing-bloom-filters.md

```diff
@@ -0,0 +1,154 @@
+```
+127.0.0.1:6379> BF.ADD ad1_filter user1
+(error) ERR operation exceeds bloom object memory limit
+```
+This poses an issue to users where their scalable bloom filters can reach the memory limit after some number of days of data population and it starts failing scale outs during the insertion of unique items.
+
+As a solution, to help users understand at what capacity their bloom filter will hit the memory limit, valkey-bloom has two options.
+These are useful to check beforehand to ensure that your bloom filter will not fail scale outs or creations later on as part of your workload.
+
+1. Perform a memory check prior to bloom filter creation
+
+We can use the `VALIDATESCALETO` option of the `BF.INSERT` command to perform a validation whether the filter is within the memory limit.
+If it is not within the limits, the command is rejected. In the example below, we see that filter1 cannot scale out and reach the capacity of 26214301 due to the memory limit. However, it can scale out and reach a capacity of 26214300.
+```
+127.0.0.1:6379> BF.INSERT filter1 VALIDATESCALETO 26214301
+(error) ERR provided VALIDATESCALETO causes bloom object to exceed memory limit
+127.0.0.1:6379> BF.INSERT filter1 VALIDATESCALETO 26214300 ITEMS item1
+1) (integer) 1
+```
+2. Check the maximum capacity that an existing scalable bloom filter can expand to
+
+We can use the `BF.INFO` command to find out the maximum capacity that the scalable bloom filter can expand to hold. In this case, we can see the filter can hold 26214300 items (after scaling out until the memory limit).
+```
+127.0.0.1:6379> BF.INFO filter1 MAXSCALEDCAPACITY
+(integer) 26214300
+```
+
+To get an idea of what the memory usage looks like for the max capacity of an individual non scaling filter, we have a table below.
+With a 128MB limit and default false positive rate, we can create a bloom filter with 112M as the capacity. With a 512MB limit, a bloom filter can hold 448M items.
+
+| Non Scaling Filter - Capacity | FP Rate | Memory Usage (MB) | Notes                                    |
+|------------------------------|---------|-------------------|------------------------------------------|
+| 112M                         | 0.01    | ~128              | Default FP Rate and Default Memory Limit |
+| 74M                          | 0.001   | ~128              | Custom FP Rate and Default Memory Limit |
+| 448M                         | 0.01    | ~512              | Default FP Rate and Custom Memory Limit |
+| 298M                         | 0.001   | ~512              | Custom FP Rate and Custom Memory Limit  |
+
+## Performance 
+
+The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of hash functions used by the bloom filter and K is the number of elements being inserted.
```

Comment:
```suggestion
The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of hash functions used by the bloom filter and K is the number of elements being inserted.
```
Earlier in the document you said K is number of hash functions, might be better to be consistent with that. You also say O(N) in the next sentence, which confused me at first since I assumed N was number of elements in the system.

Reply:
Thanks - yes, I noticed too earlier and forgot about changing this. I will update the sentence